### PR TITLE
Support for closed projects in buildship

### PR DESCRIPTION
**Pull request by fraenkelc**
### Context

Closed Gradle projects in buildships currently result in missing build dependencies. This change
adds detection for closed projects in Gradle and logic for substituting those with references to the  corresponding jars. The jars in turn are built during synchronization using a special tooling model.

This is the gradle side of https://github.com/eclipse/buildship/issues/881

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r55/RunEclipseClosedProjectTasksCrossVersionSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                .setStandardOutput(out)
+                .forTasks()
+                .run()
+        }
+
+        then:
+        def child2 = buildFinishedHandler.result.children.find { it.name == "child2" }
+        child2.projectDependencies.isEmpty()
+        child2.classpath.collect { it.file.name }.sort() == ['child1-1.0-tests.jar', 'child1-1.0.jar']
+        child2.classpath.find { it.file.name == 'child1-1.0.jar' }.classpathAttributes.find { it.name == EclipsePluginConstants.GRADLE_USED_BY_SCOPE_ATTRIBUTE_NAME }.value == "main,test"
```
Comment (donat) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r55/RunEclipseClosedProjectTasksCrossVersionSpec.groovy]: This seems unrelated to your PR

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunBuildDependenciesTaskBuilder.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    @Override
+    public Object buildAll(String modelName, EclipseRuntime eclipseRuntime, Project project) {
+        Set<String> closedProjectPaths = gatherClosedProjectPaths(eclipseRuntime);
+        Set<TaskDependency> buildDependencies = gatherBuildDependenciesForClosedProjects(closedProjectPaths);
+
+        if (!buildDependencies.isEmpty()) {
+            Project rootProject = project.getRootProject();
+            StartParameter startParameter = project.getGradle().getStartParameter();
+            List<String> taskPaths = new ArrayList<>(startParameter.getTaskNames());
+            String placeHolderTaskName = placeHolderTaskName(rootProject, "eclipseClosedProjectBuildDependencies");
```
Comment (donat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunBuildDependenciesTaskBuilder.java]: It's not a placeholder task name but rather a parent task.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunBuildDependenciesTaskBuilder.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    @Override
+    public boolean canBuild(String modelName) {
+        return "org.gradle.tooling.model.eclipse.RunClosedProjectBuildDependencies".equals(modelName);
+    }
+
+    @Override
+    public Object buildAll(String modelName, Project project) {
+        // nothing to do if no EclipseRuntime is supplied.
+        return new Object();
```
[the remark this answers is not in the document]
Comment (fraenkelc) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunBuildDependenciesTaskBuilder.java]: `ParameterizedToolingModelBuilder` is `@org.gradle.api.NonNullApi` (via `org/gradle/tooling/provider/model/package-info.java`). As such I was getting Intellij Warnings (just like in `RunEclipseTaskBuilder`):
* "not annotated method overrides method annotated with @notnull parameter" without any explicit annotations on the class
* "null' is returned by the method declared as @nonnullapi" after adding `@NonNullApi` to the class.
As a result I decided to return something, even if it is never used.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java [hunk 4]
```diff
@@ -262,6 +292,84 @@ private void populate(Project project) {
...
+                    for (File file : innerView.getFiles().getFiles()) {
+                        sourceSetUsage.put(file, sourceSet.getName().replace(",", ""));
+                    }
+                }
+
+                for (File file : artifactView.getFiles().getFiles()) {
+                    List<DefaultClasspathAttribute> attributes = createAttributes(projectDependency);
+                    Collection<String> includedInSourceSets = sourceSetUsage.get(file);
+                    if (includedInSourceSets != null) {
+                        attributes.add(new DefaultClasspathAttribute(EclipsePluginConstants.GRADLE_USED_BY_SCOPE_ATTRIBUTE_NAME, Joiner.on(',').join(includedInSourceSets)));
```
[the remark this answers is not in the document]
Comment (fraenkelc) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: Once that's merged we can remove the second ArtifactView and just match the configuration name.

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r55/RunEclipseClosedProjectTasksCrossVersionSpec.groovy (hunk 1 above)
Comment (fraenkelc) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r55/RunEclipseClosedProjectTasksCrossVersionSpec.groovy]: This verifies the correct sourceSet usage calculation in EclipseModelBuilder

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java [hunk 5]
```diff
@@ -262,6 +292,84 @@ private void populate(Project project) {
...
+    }
+
+    private List<DefaultEclipseExternalDependency> createSubstituteDependencies(Project project, ProjectDependency projectDependency) {
+        List<DefaultEclipseExternalDependency> dependencies = new ArrayList<>();
+
+        String otherProjectName = closedProjectPaths.get(StringUtils.removeStart(projectDependency.getPath(), "/"));
+        if (otherProjectName != null) {
+            // we will get duplicate dependencies from the different configurations. We use this map to deduplicate them by their respective file.
+            Map<File, DefaultEclipseExternalDependency> dependencyMap = new HashMap<>();
+            // TODO: do we need to handle minusConfigurations here?
```
[the remark this answers is not in the document]
Comment (fraenkelc) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: So, turns out handling minusConfigurations in this case  is redundant - the list of projectDependencies in populate() is already cleaned up (does not contain any projects mentioned in minusConfigurations). As such `createSubstituteDependencies` is never called for those.

**Comment by donat** [comment]
@fraenkelc can you please fix the DCO errors ☝️?

**Comment by fraenkelc** [comment]
Sorry, guess I missed that one. I'll fix it this evening.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java [hunk 6]
```diff
@@ -262,6 +292,99 @@ private void populate(Project project) {
         }
     }
 
+    private Map<String, String> gatherClosedProjectPaths(Set<String> closedProjectNames) {
+        Map<String, String> closedProjectPaths = new HashMap<>();
+        for (ProjectState state : projectRegistry.getAllProjects()) {
+            ProjectInternal project = state.getOwner().getLoadedSettings().getGradle().getRootProject().project(state.getComponentIdentifier().getProjectPath());
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: ❌ Use `IdeArtifactRegistry` instead of reaching into other project's state. Add more information to it if necessary.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java [hunk 7]
```diff
@@ -262,6 +292,99 @@ private void populate(Project project) {
...
+            EclipseModel model = project.getExtensions().getByType(EclipseModel.class);
+            if (closedProjectNames.contains(model.getProject().getName())) {
+                closedProjectPaths.put(model.getProject().getName(), project.getIdentityPath().getPath());
+            }
+        }
+        return closedProjectPaths;
+
+    }
+
+    private List<DefaultEclipseExternalDependency> createSubstituteDependencies(Project project, ProjectDependency projectDependency) {
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: ❌ If we need additional information from the dependency resolver, like what artifacts a project dependency can be replaced by, we should add it [there](https://github.com/gradle/gradle/blob/master/subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/EclipseDependenciesCreator.java#L100), not reimplement it here.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunBuildDependenciesTaskBuilder.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@NonNullApi
+public class RunBuildDependenciesTaskBuilder implements ParameterizedToolingModelBuilder<EclipseRuntime> {
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunBuildDependenciesTaskBuilder.java]: ❌ We should remove the duplication between this and the EclipseModelBuilder

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunBuildDependenciesTaskBuilder.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            String parentTaskName = parentTaskName(rootProject, "eclipseClosedDependencies");
+            Task task = rootProject.task(parentTaskName);
+            task.dependsOn(buildDependencies.toArray(new Object[0]));
+            taskPaths.add(parentTaskName);
+            startParameter.setTaskNames(taskPaths);
+        }
+        return new Object();
+    }
+
+    private Set<Task> gatherBuildDependenciesForClosedProjects(Set<String> closedProjectPaths) {
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/RunBuildDependenciesTaskBuilder.java]: ❌ Same as in the EclipseModelBuilder, add the necessary information ("buildDependencies") to the `ProjectDependency` in the Eclipse dependency resolver instead of reimplementing resolution here.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java (hunk 7 above)
Comment (fraenkelc) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: so, something along the lines of what I tried in my previous attempt? 
(see https://github.com/gradle/gradle/compare/master...fraenkelc:project_dependency_substitution?diff=unified)
What I did there was:
* extend ProjectDependency with attributes for the publication(/file) and the build dependencies (+ I had to fix the equals as otherwise I would not get multiple dependencies for different artifacts from the same project for differnt configurations)
* extract that information  from the `ComponentArtifactMetadata` in `EclipseDependenciesCreator` and pass it along to the `ProjectDependencyBuilder`
* use that information in `EclipseModelBuilder`

Comment (oehme) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: Yes, that looks a lot better 👍 

Just curious, you said this was your previous attempt: What made you drop that attempt? I didn't follow the whole history of this PR, only the final version.

Comment (fraenkelc) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: There was no technical reason(I haven't tried composites with that branch yet though) , probably just a misunderstanding on my part. I thought https://github.com/eclipse/buildship/issues/881 was asking for a solution that's local to EclipseModelBuilder and that does it's own artifact views

**Comment by fraenkelc** [comment]
I've changed the implementation over to the one in the other branch I mentioned. This version still fails a number of tests but I wanted to get it out so I can get some feedback if this is the way to move forward.

**Comment by fraenkelc** [comment]
This should be ready for another review.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/IdeDependencySet.java [hunk 10]
```diff
@@ -128,6 +128,8 @@ private ArtifactCollection getResolvedArtifacts(Configuration configuration, fin
                 public void execute(ArtifactView.ViewConfiguration viewConfiguration) {
                     viewConfiguration.lenient(true);
                     viewConfiguration.componentFilter(getComponentFilter(visitor));
+                    // TODO: Doing the restriction this way breaks a lot of test. The resulting start containing RUNTIME + TEST instead of COMPILE scopes for Idea for example.
+                    //viewConfiguration.attributes(attrs -> attrs.attribute(Usage.USAGE_ATTRIBUTE, NamedObjectInstantiator.INSTANCE.named(Usage.class, Usage.JAVA_RUNTIME_JARS)));
```
Comment (fraenkelc) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/IdeDependencySet.java]: Is there a reason why this breaks so many tests? Also is this change (COMPILE -> RUNTIME + TEST) limited to the tests or is this a breaking change for idea users? Since I'm not sure of the above I went ahead and did the filtering  in the `EclipseDependenciesCreator`

File: subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderDependenciesTest.groovy [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+            implementation child2.dependencies.project(path: ":child1")
+            testImplementation child2.dependencies.project(path: ":child1", configuration: "testArtifacts")
+        }
+        child3.eclipse {
+            (classpath as EclipseClasspath).file.whenMerged {
+                def customDependency = new ProjectDependency("/child1")
+                customDependency.buildDependencies(":child1:jar")
+                customDependency.publication = fileReference("customJar.jar")
+                customDependency.publicationJavadocPath = fileReference("customJar-javadoc.jar")
+                customDependency.publicationSourcePath = fileReference("customJar-sources.jar")
```
Comment (fraenkelc) [inline subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderDependenciesTest.groovy]: This is ugly, but it's at least some way to specify javadoc and sources for the replacements. I haven't seen an easy way to get javadoc and sources artifacts from gradle - especially considering that these artifacts may be from non-default configurations.
I'm missing a good way to get from the `ProjectDependency` to the gradle `Project` here. Is there one / would it make sense to introduce that as a property on the `ProjectDependency`?

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/IdeDependencySet.java (hunk 10 above)
Comment (oehme) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/IdeDependencySet.java]: You can't limit to runtime jars here, since compile dependencies are not jars, but classes folders.

File: subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderDependenciesTest.groovy (hunk 11 above)
Comment (oehme) [reply subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderDependenciesTest.groovy]: > I'm missing a good way to get from the ProjectDependency to the gradle Project here. 

This is deliberate, you shouldn't be able to reach into projects through the resolution engine, projects should be isolated from each other. If we really wanted the javadoc/sources, we should make them variants of the project that we can query for, just like RUNTIME_JARS.

I'd leave this out for now and maybe give it a shot in a different PR.

**Comment by oehme** [comment]
This looks great, thanks for changing it.

> This version still fails a number of tests 

Anything I can help you with?

**Comment by fraenkelc** [comment]
I had some flaky tests in unrelated code but I blame those on my slow windows dev machine. I can go ahead and remove the two comments regarding the variant filtering. After that I think I'm done with the PR unless there's other reviews feedback to incorporate. I'd be interested in seeing another teamcity build of this, the last one (using the previous implementation) had a strange test failure.

**Comment by oehme** [comment]
I'll run the test suite now.

**Comment by fraenkelc** [comment]
And that's the changes I was talking about done. How do we move forward with this pr?

**Comment by oehme** [comment]
I'm just waiting for a green build, including our cross-version tests for the tooling API. After that this is good to go :)

**Comment by donat** [comment]
The build successfully finished:
https://builds.gradle.org/viewLog.html?buildId=23328236&tab=buildResultsDiv&buildTypeId=Gradle_Check_Stage_ReadyforMerge_Trigger&branch_Gradle_Check=pull%2F9405%2Fhead.
Shall I proceed with the merge?
Donát Csikós

Senior Software Engineer
Gradle
W. gradle.com

**Comment by oehme** [comment]
Yes please go ahead. Thanks again @fraenkelc for this excellent contribution!

**Comment by fraenkelc** [comment]
nice, once there's a nightly for it I can work on a proper pull request for the Buildship side :)
